### PR TITLE
Issue: #791 Add args and check condition for remote GPU

### PR DESCRIPTION
**Pull request by Zhiyuan-w**
# Description

Please include a summary of the change and which issue is fixed. Please also include
relevant motivation and context. List any dependencies that are required for this
change.

Fixes #<issue_number>

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ v] Bugfix (non-breaking change which fixes an issue)
- [v ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [v ] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [v ] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ v] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by Zhiyuan-w** [comment]
> Thanks for proposed fix.
> 
> I think we should make the gpu arg in the CLI optional. The current fix is a bit confusing - requires an extra argument "remote" and/or using "all" in the gpu argument where usually the arguments have been string numbers.
> 
> How about:
> 
> ```python
> if gpu:
>         os.environ["CUDA_VISIBLE_DEVICES"] = gpu        
>        os.environ["TF_FORCE_GPU_ALLOW_GROWTH"] = "true" if gpu_allow_growth else "false"
> ```
> 
> and we make
> 
> ```python
>     gpu: str=None in train, predict
> ```
> 
> and
> 
> ```python
>  "--gpu",        "-g",        help="GPU index for training."        '-g "" for using CPU'        '-g "0" for using GPU 0'        '-g "0,1" for using GPU 0 and 1.',        type=str,        **required=False**,  default=None  )
> ```

Sure.

**Comment by Zhiyuan-w** [comment]
> Thanks for proposed fix.
> 
> I think we should make the gpu arg in the CLI optional. The current fix is a bit confusing - requires an extra argument "remote" and/or using "all" in the gpu argument where usually the arguments have been string numbers.
> 
> How about:
> 
> ```python
> if gpu:
>         os.environ["CUDA_VISIBLE_DEVICES"] = gpu        
>        os.environ["TF_FORCE_GPU_ALLOW_GROWTH"] = "true" if gpu_allow_growth else "false"
> ```
> 
> and we make
> 
> ```python
>     gpu: str=None in train, predict
> ```
> 
> and
> 
> ```python
>  "--gpu",        "-g",        help="GPU index for training."        '-g "" for using CPU'        '-g "0" for using GPU 0'        '-g "0,1" for using GPU 0 and 1.',        type=str,        **required=False**,  default=None  )
> ```

It may cause a SyntaxError: non-default argument follows default argument. As I do:
```
    gpu: str=None in train, predict
```
 Could I keep the "--gpu" as required and add a "--remote"  arg? Like
```
"--remote",        "-r",          help="Use GPU remotely for training.",           dest="remote",          action="store_true",  
     required=False,       default=False
```
In train()
```
train(....., remote:bool = False)
if not remote:
       os.environ["CUDA_VISIBLE_DEVICES"] = gpu        
       os.environ["TF_FORCE_GPU_ALLOW_GROWTH"] = "true" if gpu_allow_growth else "false"
```

**Comment by Zhiyuan-w** [comment]
> It may cause a SyntaxError: non-default argument follows default argument. As I do:
> 
> ```
>     gpu: str=None in train, predict
> ```
> 
> Could I keep the "--gpu" as required and add a "--remote" arg? Like
> 
> ```
> "--remote",        "-r",          help="Use GPU remotely for training.",           dest="remote",          action="store_true",  
>      required=False,       default=False
> ```
> 
> In train()
> 
> ```
> train(....., remote:bool = False)
> if not remote:
>        os.environ["CUDA_VISIBLE_DEVICES"] = gpu        
>        os.environ["TF_FORCE_GPU_ALLOW_GROWTH"] = "true" if gpu_allow_growth else "false"
> ```

My wifi is not stable right now. Sorry for sending comments twice.

**Comment by NMontanaBrown** [comment]
@Zhiyuan-w could you have a look at the comments? I think there are some users waiting for this fix at the moment so it would be great to close this PR sooner rather than later!

**Comment by NMontanaBrown** [comment]
@Zhiyuan-w ok so instead of making gpu:str = None in the functions, just make the default arg in GPU CLI None or False. That way, the same logic applies but we don't need to change the order of the args or add remote to the CLI. So simply passing "GPU" arg to CLI will trigger CUDA_VISIBLE_DEVICES, if it is not passed it will be put in as "False". Does that make sense?

**Comment by Zhiyuan-w** [comment]
Good idea. I will push the new commit today.

File: deepreg/predict.py [hunk 1]
```diff
@@ -321,11 +324,13 @@ def main(args=None):
...
         "-g",
         help="GPU index for training."
+        "-g for using GPU remotely"
         '-g "" for using CPU'
         '-g "0" for using GPU 0'
         '-g "0,1" for using GPU 0 and 1.',
         type=str,
-        required=True,
+        required=False,
+        default=False,
```
Comment (mathpluscode) [inline deepreg/predict.py:333]: this arg is of type str, default should not be boolean?

I wonder, if we do not make it required, and there's no default, what will be the value?

File: demos/grouped_mask_prostate_longitudinal/demo_predict.py [hunk 2]
```diff
@@ -48,8 +49,8 @@
     config_path.append("config/test/demo_unpaired_grouped.yaml")
 
 predict(
-    gpu="0",
-    gpu_allow_growth=True,
+    gpu="",
```
Comment (mathpluscode) [inline demos/grouped_mask_prostate_longitudinal/demo_predict.py:52]: why changing this, it should no be linked to this PR?

File: deepreg/predict.py (hunk 1 above)
Comment (Zhiyuan-w) [reply deepreg/predict.py:333]: If there is no default， it gives None.
I tried the default can be bool False，
as the type is str.

File: demos/grouped_mask_prostate_longitudinal/demo_predict.py (hunk 2 above)
Comment (Zhiyuan-w) [reply demos/grouped_mask_prostate_longitudinal/demo_predict.py:52]: yerh…it should be my mistake.

File: deepreg/predict.py (hunk 1 above)
Comment (Zhiyuan-w) [reply deepreg/predict.py:333]: Should I just remove the default value？
If None can gives the same effect as False.

File: demos/grouped_mask_prostate_longitudinal/demo_predict.py [hunk 3]
```diff
@@ -36,6 +36,7 @@
     f"--log_dir demos/{name} "
     "--exp_name logs_predict "
     "--save_png --split test\n"
+    "If using remote GPU, add '--remote'\n"
```
Comment (mathpluscode) [inline demos/grouped_mask_prostate_longitudinal/demo_predict.py:39]: Why adding this line?

File: demos/grouped_mask_prostate_longitudinal/demo_train.py [hunk 4]
```diff
@@ -33,6 +34,7 @@
     f"--config_path demos/{name}/{name}.yaml "
     f"--log_dir demos/{name} "
     "--exp_name logs_train\n"
+    "If using remote GPU, add '--remote' \n"
```
Comment (mathpluscode) [inline demos/grouped_mask_prostate_longitudinal/demo_train.py:37]: Same here

**Review by mathpluscode — APPROVED** [review]
LGTM

**Comment by NMontanaBrown** [comment]
Changelog needs to be updated before merge.

**Review by NMontanaBrown — APPROVED** [review]

**Comment by NMontanaBrown** [comment]
@mathpluscode @YipengHu can you review please?

**Comment by NMontanaBrown** [comment]
@Zhiyuan-w Can you merge main into branch so that it is up to date?

**Comment by NMontanaBrown** [comment]
> @Zhiyuan-w Can you merge main into branch so that it is up to date?

BUMP

**Comment by NMontanaBrown** [comment]
Thanks for your contribution! @Zhiyuan-w , also @alkististav, check the fix!

**Pull request merged**